src/mutex.cpp

```diff
@@ -4,51 +4,55 @@
 #include "syscallwrappers.h"
 
 
+static const uint32_t LOCK_FREE = 0;
+static const uint32_t LOCK_ACQUIRED = 1;
+static const uint32_t LOCK_ACQUIRED_WAITERS_QUEUED = 2;
+
 /*
- * Mutex
+ * Mutex has following fields:
+ *   futex: The field for all futex operations
+ *   type:  Lock type.
+ *   owner: Used for recursive locks and to check for deadlocks.
+ *   count: Record recursive lock acquisition count.
  */
 
 extern "C" void
 DmtcpMutexInit(DmtcpMutex *mutex, DmtcpMutexType type)
 {
   mutex->type = type;
+  mutex->futex = 0;
   mutex->owner = 0;
   mutex->count = 0;
 }
 
 
+ /*
+ * If futex is 0 (not acquired), set to 1 (acquired with no waiters) and return.
+ * Otherwise, ensure that it is >1 (acquired, possibly with waiters) and then
+ * block until we acquire the lock, at which point futex will still be >1.  The
+ * lock is always acquired on return.
+ */
 extern "C" int
 DmtcpMutexLock(DmtcpMutex *mutex)
 {
-  pid_t owner = 1;
-
-  if (mutex->type != DMTCP_MUTEX_LLL) {
-    owner = dmtcp_gettid();
-
-    if (mutex->owner == owner) {
-      if (mutex->type == DMTCP_MUTEX_RECURSIVE) {
-        JASSERT(mutex->count + 1 != 0);
-        mutex->count++;
-        return 0;
-      }
-      return EDEADLK;
-    }
+  if (DmtcpMutexTryLock(mutex) == 0) {
+    return 0;
   }
 
-  while (1) {
-    uint32_t waitVal =
-      __sync_val_compare_and_swap((uint32_t *)&mutex->owner, 0, owner);
-
-    if (waitVal == 0) {
-      // We successfully acquired the lock.
-      break;
+  do {
```

Comment:
A comment above the `do` would help.  Readers can figure out the code, but it takes time.  It's better to show the intention in the comment, and a reader will either temporarily trust that it's correct, or will validate against the spec in the comment.

I think the intention here is:
```
// If someone else is holding the lock (lock in state LOCK_ACQUIRED), then we register ourselves
// for waiting on the lock by putting it in state LOCK_ACQUIRED_WAITERS_QUEUED, and then
// calling futex_wait().
// When we wake up from this, we test if the lock is in state LOCK_FREE, and if so, we set it for LOCK_ACQUIRED_WAITERS_QUEUED, and we return, holding this lock.
// If the lock was originally in state LOCK_FREE, then we set the lock to state LOCK_ACQUIRED_WAITERS_QUEUED.
// We succeed and return only by leaving the lock in state LOCK_ACQUIRED_WAITERS_QUEUED.  Else, we loop.
```
QUESTION:  Suppose we are the only one who acquired the lock.  Shouldn't we then be allowed to return by leaving the lock in state LOCK_ACQUIRED?

Reply:
Fixed.